pkg/prom/instance/instance.go

```diff
@@ -138,15 +122,44 @@ func (c *Config) ApplyDefaults(global *config.GlobalConfig) {
 			}
 		}
 
+		if _, ok := jobNames[sc.JobName]; ok {
+			return fmt.Errorf("found multiple scrape configs with job name %q", sc.JobName)
+		}
+		jobNames[sc.JobName] = struct{}{}
+
 		sc.RelabelConfigs = append(sc.RelabelConfigs, DefaultRelabelConfigs...)
 	}
-}
 
-// Validate checks if the Config has all required fields filled out.
-// This should only be called after ApplyDefaults.
-func (c *Config) Validate() error {
-	if c.Name == "" {
-		return errors.New("missing instance name")
+	rwNames := map[string]struct{}{}
+	for _, cfg := range c.RemoteWrite {
+		if cfg == nil {
+			return fmt.Errorf("empty or null remote write config section")
+		}
+
+		// Typically Prometheus ignores empty names here, but we need to assign a
+		// unique name to the config so we can pull metrics from it when running
+		// an instance.
+		var generatedName bool
+		if cfg.Name == "" {
+			hash, err := getHash(cfg)
+			if err != nil {
+				return err
+			}
+
+			// We have to add the name of the instance to ensure that generated metrics
+			// are unique across multiple agent instances. The remote write queues currently
+			// globally register their metrics so we can't inject labels here.
+			cfg.Name = c.Name + "-" + hash[:6]
+			generatedName = true
+		}
+
+		if _, ok := rwNames[cfg.Name]; ok {
```

Comment:
tiny nit: `exists` or `found` instead of `ok`. It's perfectly understandable just kind of weird to have `ok` result in an error :man_shrugging: